# **🚀 Факторизация и другие интересные операции в Julia**  
# **Урок Тринадцатый.**

По мотивам работы Андреаса Ноака Йенсена (MIT & JuliaComputing) [(Twitter)](https://x.com/anoackjensen)  
(с авторскими дополнениями Сергея Соболевского)

<br>

##  **📌 Темы:**  

1. **Что такое факторизация матрицы**
   
2. **Основные виды факторизации матриц.**
   
   2.1 **LU-факторизация**

   2.2 **QR-факторизация**

   2.3 **Cholesky-факторизация**

   2.4 **Сингулярное разложение (SVD)**

   2.5 **Разложение по собственным значениям и собственным векторам (EVD)**

   
3. **Специальные матричные структуры.**
   
4. **Общая линейная алгебра.**



<br>


<br>


<br>

---

<br>

# 📚 **1. Что такое факторизация матрицы?**

**Факторизация матрицы** (или разложение матрицы) — это представление исходной матрицы в виде произведения нескольких матриц определённого типа, обычно более простых по структуре. 

Суть метода — исходную задачу (например, решение системы линейных уравнений, нахождение определителя или обратной матрицы) привести к более простой и численно устойчивой форме.

<br>


### 🧮 Для чего нужна факторизация?

Факторизация применяется в различных областях математики и вычислительной науки:

- **Решение систем линейных уравнений**: упрощает решение, особенно при множестве задач с одинаковой матрицей, но разными правыми частями.
- **Нахождение обратных матриц**: облегчает и ускоряет процесс вычисления.
- **Определение ранга матрицы и её структуры** (например, собственных значений и векторов).
- **Упрощение вычислений** в численных методах, повышая устойчивость и точность расчётов.

<br>



<br>

<br>


### ⚙️ Настройка нашей среды **Julia**.

Прежде чем начать, давайте настроим нашу среду , а заодно повторим как подключить в **Julia** библиотеку `LinearAlgebra`, чтобы дальше мы могли применять факторизации и работать со специальными структурами матриц.  

- **Подключаем пакет `LinearAlgebra` :**

In [1]:
using LinearAlgebra # Подключаем пакет LinearAlgebra



- **Создаем случайную матрицу $A$ (3x3) :**

In [2]:
A = rand(3, 3)  # Создаем случайную матрицу 3x3

3×3 Matrix{Float64}:
 0.472011  0.833188   0.516267
 0.926891  0.0625341  0.590671
 0.444311  0.778406   0.55915

- **Создаем вектор $x$ из трех единиц :**

In [3]:
x = fill(1, (3,))  # Создаем вектор из трех 1-единиц

3-element Vector{Int64}:
 1
 1
 1

- **Вычисляем для пробы результат умножения матрицы $A$ на вектор $x$ :**

In [4]:
b = A * x  # Вычисляем результат умножения матрицы A на вектор x

3-element Vector{Float64}:
 1.821466743264216
 1.58009585183461
 1.7818669244725784

**🚩 "Бинго!" Всё получилось, идём дальше!**

<br>

---

<br>

## **🔢 2. Основные виды факторизации матриц.**

<br>
Наиболее часто применяются следующие виды разложений:

1. **LU-факторизация**
   
2. **QR-факторизация**
   
3. **Cholesky-факторизация**
   
4. **Сингулярное разложение (SVD)**
   
5. **Разложение по собственным значениям и собственным векторам (EVD)**

<br>

Давайте рассмотрим кратко каждый этот тип подробнее, но вначале настроим нашу среду **Julia** для работы.

<br>

---

<br>

### 🟢 **2.1 LU-факторизация**

<br>

**LU-факторизация** — это представление матрицы $A$ в виде произведения нижней треугольной матрицы $L$ (Lower triangular) и верхней треугольной матрицы $U$ (Upper triangular):

$$
A = L \cdot U
$$

Чтобы повысить численную устойчивость вычислений, обычно используется матрица перестановок $P$:

$$
P \cdot A = L \cdot U
$$

Здесь:

- $P$ — матрица перестановок (переставляет строки исходной матрицы $A$),
- $L$ — нижняя треугольная матрица с единичными элементами на главной диагонали,
- $U$ — верхняя треугольная матрица.

<br>

#### 🔹 **Применение LU-факторизации:**

- Решение систем линейных уравнений.
- Вычисление определителей матриц.
- Нахождение обратной матрицы.


В языке Julia LU-факторизация матрицы $A$ выполняется с помощью функции `lu()`:

In [5]:
Alu = lu(A)  # Разложение матрицы A на L и U

LU{Float64, Matrix{Float64}, Vector{Int64}}
L factor:
3×3 Matrix{Float64}:
 1.0       0.0       0.0
 0.509242  1.0       0.0
 0.479357  0.933969  1.0
U factor:
3×3 Matrix{Float64}:
 0.926891  0.0625341  0.590671
 0.0       0.801343   0.215473
 0.0       0.0        0.0747623

👉 Julia автоматически создаёт объект специального составного типа, который хранит компоненты факторизации.

Извлечь компоненты факторизации можно с помощью специальных полей объекта:

- **Матрица перестановок:**

In [6]:
P = Alu.P

3×3 Matrix{Float64}:
 0.0  1.0  0.0
 1.0  0.0  0.0
 0.0  0.0  1.0

- **Нижняя треугольная матрица $L$ :**

In [7]:
L = Alu.L

3×3 Matrix{Float64}:
 1.0       0.0       0.0
 0.509242  1.0       0.0
 0.479357  0.933969  1.0

- **Верхняя треугольная матрица $U$ :**

In [8]:
U = Alu.U

3×3 Matrix{Float64}:
 0.926891  0.0625341  0.590671
 0.0       0.801343   0.215473
 0.0       0.0        0.0747623

### 🗒 **Примеры использования LU-факторизации:**

#### 🔹 **Пример 1: Решение системы линейных уравнений:**

Пусть задана система линейных уравнений:

$$
\begin{cases}
2x + 3y = 5 \\
4x + 5y = 11
\end{cases}
$$

Запишем в матричной форме:

In [9]:
A = [2 3; 4 5]
b = [5; 11]

2-element Vector{Int64}:
  5
 11

In [10]:
Alu = lu(A) # Разложение матрицы A на L и U

LU{Float64, Matrix{Float64}, Vector{Int64}}
L factor:
2×2 Matrix{Float64}:
 1.0  0.0
 0.5  1.0
U factor:
2×2 Matrix{Float64}:
 4.0  5.0
 0.0  0.5

- **Решение через исходную матрицу:**

In [11]:
x_direct = A \ b  # Решение системы уравнений Ax = b с помощью обратного хода

2-element Vector{Float64}:
  4.0
 -1.0

- **Решение через факторизацию LU:**

In [12]:
x_lu = Alu \ b  # Решение системы уравнений Ax = b с помощью LU-разложения

2-element Vector{Float64}:
  4.0
 -1.0

**Убеждаемся, что оба метода дают одинаковый результат.**


#### 🔹 **Пример 2: Вычисление определителя:**

Определитель матрицы можно вычислить напрямую или используя LU-факторизацию:

In [13]:
det_A_direct = det(A) # Вычисление определителя матрицы A с помощью обратного хода

-2.0

In [14]:

det_A_lu = det(Alu) # Вычисление определителя матрицы A с помощью LU-разложения

-2.0

Результаты будут близки до численной погрешности:

In [15]:
det(A) ≈ det(Alu) # Проверка равенства определителей

true

#### 👉 **Преимущества LU-факторизации в Julia:**

- **Численная стабильность**: Благодаря матрице перестановок факторизация становится устойчивой.
  
- **Эффективность**: LU-факторизацию можно использовать повторно для различных вычислений (решение систем, нахождение обратной матрицы, вычисление определителей).



<br>

---

<br>


<br>

### 🟢 **2.2 QR-факторизация**

**QR-факторизация** — это разложение матрицы $ A $ в виде произведения двух специальных матриц: ортогональной (унитарной) матрицы $ Q $ и верхней треугольной матрицы $ R $ :

<br>

$$
A = Q \cdot R
$$

**Характеристики матриц разложения:**

- Матрица $ Q $ — ортогональная (унитарная), то есть выполняется:
$$
Q^T Q = I
$$

Это означает, что столбцы матрицы $ Q $ :

- **ортогональны** друг другу (их скалярное произведение равно нулю);
- имеют **единичную длину** (нормированы).

- Матрица $ R $ — верхняя треугольная, то есть все её элементы ниже главной диагонали равны нулю.


<br>

#### 📌 **Применение QR-факторизации:**

QR-факторизация широко используется в вычислительной математике, статистике и инженерии:

- **Решение задач наименьших квадратов** (линейная регрессия).
  
- **Решение линейных систем уравнений** (особенно устойчивых к численным погрешностям).
  
- **Вычисление собственных значений и векторов** (QR-алгоритм).

<br>

### 🗒  **Примеры выполнения QR-факторизации в Julia:**

#### 🔹 **Пример 1.**

В языке Julia **QR-факторизацию** матрицы $ A $ можно выполнить с помощью функции `qr()`:


In [16]:
Aqr = qr(A) # Разложение матрицы A на Q и R

LinearAlgebra.QRCompactWY{Float64, Matrix{Float64}, Matrix{Float64}}
Q factor: 2×2 LinearAlgebra.QRCompactWYQ{Float64, Matrix{Float64}, Matrix{Float64}}
R factor:
2×2 Matrix{Float64}:
 -4.47214  -5.81378
  0.0      -0.447214


После этого результат факторизации сохраняется в специальном объекте, из которого легко извлечь матрицы $ Q $ и $ R $ :

- **Матрица $ Q $** (ортогональная матрица):

In [17]:
Q = Aqr.Q

2×2 LinearAlgebra.QRCompactWYQ{Float64, Matrix{Float64}, Matrix{Float64}}

- **Матрица $ R $** (верхняя треугольная матрица):

In [18]:
R = Aqr.R

2×2 Matrix{Float64}:
 -4.47214  -5.81378
  0.0      -0.447214

👉 Таким образом, после выполнения **QR-факторизации** мы можем работать с её компонентами напрямую и эффективно.

#### 🔹 **Пример 2.**

**Краткий пример использования, :) для закрепления темы:**

Рассмотрим простую матрицу:

```julia
A = [1 2; 
     3 4; 
     5 6]
```

In [19]:
A = [1 2; 3 4; 5 6]

3×2 Matrix{Int64}:
 1  2
 3  4
 5  6

In [20]:
Aqr = qr(A) # Разложение матрицы A на Q и R

LinearAlgebra.QRCompactWY{Float64, Matrix{Float64}, Matrix{Float64}}
Q factor: 3×3 LinearAlgebra.QRCompactWYQ{Float64, Matrix{Float64}, Matrix{Float64}}
R factor:
2×2 Matrix{Float64}:
 -5.91608  -7.43736
  0.0       0.828079

In [21]:
Q = Aqr.Q   # ортогональная матрица

3×3 LinearAlgebra.QRCompactWYQ{Float64, Matrix{Float64}, Matrix{Float64}}

In [22]:
R = Aqr.R   # верхняя треугольная матрица

2×2 Matrix{Float64}:
 -5.91608  -7.43736
  0.0       0.828079

👉 **Теперь исходная матрица $ A $ может быть представлена в виде:**

$$
A = Q \cdot R
$$


In [23]:
A = Q * R  # Проверка разложения

3×2 Matrix{Float64}:
 1.0  2.0
 3.0  4.0
 5.0  6.0

<br>

#### 👉 **Преимущества использования QR-факторизации:**

- **Численная устойчивость**: QR-факторизация обычно более стабильна и точна, чем другие методы (например, LU-факторизация), особенно при работе с плохо обусловленными матрицами.
  
- **Эффективность**: Позволяет быстро и устойчиво решать системы линейных уравнений и задачи регрессии.



<br>

---

<br>

<br>

### 🟢 **2.3 Cholesky-факторизация**

**Cholesky-факторизация** применяется исключительно для **симметричных и положительно определённых** матриц. Она представляется в виде произведения нижней треугольной матрицы $L$ и её транспонированной матрицы $L^T$:

$$
A = L \cdot L^T
$$

где:

- $L$ — нижняя треугольная матрица (все элементы выше диагонали равны нулю).
- $L^T$ — верхняя треугольная матрица, транспонированная к $L$.

<br>

##### В Julia **Cholesky-факторизация** выполняется функцией `cholesky()` .

<br>

### 🔹 **Особенности Cholesky-факторизации:**

- Используется только для симметричных и положительно определённых матриц.
- Является более простым и численно стабильным методом по сравнению с LU-факторизацией.
- Более эффективна с точки зрения производительности, так как требует примерно в два раза меньше вычислений, чем LU-факторизация.

<br>


#### 📌 **Применение Cholesky-факторизации:**

- Решение систем линейных уравнений (особенно с симметричными матрицами).
  
- Вычисления в статистике (например, линейные регрессионные модели, многомерное нормальное распределение).
  
- Численные методы оптимизации и анализа данных.

<br>


### 🗒  **Примеры Cholesky-факторизации в Julia:**

#### 🔹 **Пример 1.**

1. **Определяем симметричную положительно определённую матрицу $A$:**

```julia
A = [
    25.0  15.0  -5.0;
    15.0  18.0   0.0;
    -5.0   0.0  11.0
]
```


In [24]:
A = [
    25.0  15.0  -5.0;
    15.0  18.0   0.0;
    -5.0   0.0  11.0
]

3×3 Matrix{Float64}:
 25.0  15.0  -5.0
 15.0  18.0   0.0
 -5.0   0.0  11.0

2. **Выполняем проверку симметричности :**

In [25]:
println("Is A symmetric? ", issymmetric(A)) # Проверка симметричности матрицы A

Is A symmetric? true

3. **Выполнение Cholesky-факторизации :**

In [26]:
cholesky_factor = cholesky(A) # Разложение матрицы A на L и L^T

Cholesky{Float64, Matrix{Float64}}
U factor:
3×3 UpperTriangular{Float64, Matrix{Float64}}:
 5.0  3.0  -1.0
  ⋅   3.0   1.0
  ⋅    ⋅    3.0

4. **Выводим результат, т.е. извлекаем компоненты факторизации следующим образом:**

- **Нижняя треугольная матрица $L$ :**

In [27]:
L = cholesky_factor.L # нижняя треугольная матрица

3×3 LowerTriangular{Float64, Matrix{Float64}}:
  5.0   ⋅    ⋅ 
  3.0  3.0   ⋅ 
 -1.0  1.0  3.0

In [28]:
println("Cholesky factor L:\n", cholesky_factor.L) # Матрица L

Cholesky factor L:


[5.0 0.0 0.0; 3.0 3.0 0.0; -1.0 1.0 3.0]


#### 🔹 **Пример 2. Использования Cholesky-факторизации, ещё один:** :*

Рассмотрим симметричную и положительно определённую матрицу:

In [29]:
A = [4 2; 2 3]

2×2 Matrix{Int64}:
 4  2
 2  3

- **Проверяем положительную определённость (собственные значения положительны):**

In [30]:
eigvals(A) # Вычисление собственных значений матрицы A

2-element Vector{Float64}:
 1.43844718719117
 5.561552812808831


- **Выполняем Cholesky-факторизацию:**

In [31]:
A_chol = cholesky(A)

Cholesky{Float64, Matrix{Float64}}
U factor:
2×2 UpperTriangular{Float64, Matrix{Float64}}:
 2.0  1.0
  ⋅   1.41421

- **Извлекаем нижнюю треугольную матрицу L :**

In [32]:
L = A_chol.L

2×2 LowerTriangular{Float64, Matrix{Float64}}:
 2.0   ⋅ 
 1.0  1.41421

- **Проверим правильность разложения :**

In [33]:
A_reconstructed = L * L' # Восстановление матрицы A

2×2 Matrix{Float64}:
 4.0  2.0
 2.0  3.0

- **Проверка равенства определителей :**

In [34]:
det(A_reconstructed) ≈ det(A) # Проверка равенства определителей

true

#### 👉 **Преимущества Cholesky-факторизации:**

- Высокая численная устойчивость.
  
- Минимальные требования к памяти.
  
- Повышенная производительность для специальных типов матриц.

<br>

🚀 Используя **Cholesky-факторизацию** в Julia, можно эффективно и стабильно решать системы линейных уравнений, часто возникающие в прикладных задачах статистики и численного анализа.

<br>

---

<br>

### 🟢 **2.4 Сингулярное разложение (SVD, Singular Value Decomposition)**

**Сингулярное разложение (SVD)** — это факторизация любой матрицы $A$ (включая неквадратные) в виде произведения трёх матриц:

<br>

$$
A = U \cdot \Sigma \cdot V^T
$$

**где:**

- $U$ — ортогональная матрица размера $m \times m$, содержащая левые сингулярные векторы.
  
- $V$ — ортогональная матрица размера $n \times n$, содержащая правые сингулярные векторы.
  
- $\Sigma$ — прямоугольная диагональная матрица размера $m \times n$, содержащая сингулярные числа на диагонали (упорядочены по убыванию и всегда неотрицательные).
  
<br>

##### **В Julia сингулярное разложение выполняется функцией `svd()`.**

<br>

#### 🔹 **Особенности сингулярного разложения (SVD):**

- Работает для любой матрицы (квадратной и неквадратной).
  
- Диагональные элементы (сингулярные числа) матрицы $\Sigma$ неотрицательны и расположены по убыванию.
  
- Позволяет оценивать ранг матрицы и её численную устойчивость.

<br>

#### 📌 **Применение SVD:**

- **Сжатие и обработка изображений:** Уменьшение размерности данных, PCA (метод главных компонент).
  
- **Анализ данных и машинное обучение:** Рекомендательные системы, кластеризация, обработка естественного языка.
  
- **Определение ранга матрицы:** Оценка численной стабильности и качества аппроксимации данных.
  
  <br>

### 🗒  **Примеры SVD-разложения в Julia:**

#### 🔹 **Пример 1.**

1. **Определяем симметричную положительно определённую матрицу $A$:**

```julia
A = [
    25.0  15.0  -5.0;
    15.0  18.0   0.0;
    -5.0   0.0  11.0
]
```

In [35]:
A = [
    25.0  15.0  -5.0;
    15.0  18.0   0.0;
    -5.0   0.0  11.0
]

3×3 Matrix{Float64}:
 25.0  15.0  -5.0
 15.0  18.0   0.0
 -5.0   0.0  11.0

2. **Выполним сингулярное разложение при помощи функцией `svd()`:**

In [36]:
Asvd = svd(A)

SVD{Float64, Float64, Matrix{Float64}, Vector{Float64}}
U factor:
3×3 Matrix{Float64}:
 -0.783736   0.178127   0.595003
 -0.603219  -0.446485  -0.660893
  0.147937  -0.876882   0.457375
singular values:
3-element Vector{Float64}:
 37.48885385630933
 12.015683645427238
  4.495462498263424
Vt factor:
3×3 Matrix{Float64}:
 -0.783736  -0.603219   0.147937
  0.178127  -0.446485  -0.876882
  0.595003  -0.660893   0.457375

3. **Полученные компоненты факторизации извлекаются следующим образом:**
   


- Матрица $U$ (левые сингулярные векторы):

In [37]:
U = Asvd.U

3×3 Matrix{Float64}:
 -0.783736   0.178127   0.595003
 -0.603219  -0.446485  -0.660893
  0.147937  -0.876882   0.457375

- Диагональная матрица сингулярных чисел $\Sigma$:

In [38]:
S = Asvd.S

3-element Vector{Float64}:
 37.48885385630933
 12.015683645427238
  4.495462498263424

- Матрица $V$ (правые сингулярные векторы):

In [39]:
V = Asvd.V

3×3 adjoint(::Matrix{Float64}) with eltype Float64:
 -0.783736   0.178127   0.595003
 -0.603219  -0.446485  -0.660893
  0.147937  -0.876882   0.457375

- **Восстановление исходной матрицы по её SVD-разложению :**

A_reconstructed = U * Diagonal(S) * V'

### 🔹 **Пример 2 использования SVD:**

Рассмотрим новую матрицу:

```julia
A = [3 1; 1 3; 1 1]
```



In [40]:
A = [3 1; 1 3; 1 1]

3×2 Matrix{Int64}:
 3  1
 1  3
 1  1

- **Выполняем SVD-разложение :**

In [41]:
Asvd = svd(A) # Сингулярное разложение матрицы A

SVD{Float64, Float64, Matrix{Float64}, Vector{Float64}}
U factor:
3×2 Matrix{Float64}:
 -0.666667   0.707107
 -0.666667  -0.707107
 -0.333333   0.0
singular values:
2-element Vector{Float64}:
 4.242640687119286
 2.0
Vt factor:
2×2 Matrix{Float64}:
 -0.707107  -0.707107
  0.707107  -0.707107

- **Матрица $U$ (левые сингулярные векторы):**

In [42]:
U = Asvd.U # ортогональная матрица

3×2 Matrix{Float64}:
 -0.666667   0.707107
 -0.666667  -0.707107
 -0.333333   0.0

- **Диагональная матрица сингулярных чисел $\Sigma$ :**

In [43]:
S = Asvd.S # диагональная матрица

2-element Vector{Float64}:
 4.242640687119286
 2.0

- **Матрица $V$ (правые сингулярные векторы):**

In [44]:
V = Asvd.V # ортогональная матрица

2×2 adjoint(::Matrix{Float64}) with eltype Float64:
 -0.707107   0.707107
 -0.707107  -0.707107

- **Восстановление исходной матрицы по её SVD-разложению:**

In [45]:
A_reconstructed = U * Diagonal(S) * V' # Восстановление матрицы A

3×2 Matrix{Float64}:
 3.0  1.0
 1.0  3.0
 1.0  1.0

#### 👉 **Преимущества в использовании SVD-разложения:**

- Позволяет работать с любыми матрицами, включая неквадратные.
  
- Высокая численная устойчивость и надёжность.
  
- Эффективно решает задачи, связанные с обработкой больших объёмов данных, особенно при анализе данных и машинном обучении.

<br>

<br>

---

<br>

### 🟢 **2.5 Собственные разложения (Eigendecompositions -EVD)**

<br>

**Собственное разложение (Eigendecomposition)** — это представление её в виде произведения трёх матриц:

<br>

$$
A = V \Lambda V^{-1},
$$


**где:**

- $ V $ — матрица собственных векторов,
- $ \Lambda $ — диагональная матрица собственных значений.


### 🗒  **Примеры EVD-разложения в Julia:**

#### 🔹 **Пример 1.**

1. **Предположим, что у нас есть симметричная матрица $A$, созданная следующим образом :**


```julia
A = [
    25.0  15.0  -5.0;
    15.0  18.0   0.0;
    -5.0   0.0  11.0
]
```

In [46]:
A = [
    25.0  15.0  -5.0;
    15.0  18.0   0.0;
    -5.0   0.0  11.0
]

3×3 Matrix{Float64}:
 25.0  15.0  -5.0
 15.0  18.0   0.0
 -5.0   0.0  11.0

- **Или мы можем создать симметричную матрицу:**

In [47]:
Asym = A + A'# Создание симметричной матрицы

3×3 Matrix{Float64}:
  50.0  30.0  -10.0
  30.0  36.0    0.0
 -10.0   0.0   22.0

- **Чтобы вычислить собственное разложение, используем функцию `eigen`:**

In [48]:
AsymEig = eigen(Asym) # Вычисление собственных значений и векторов матрицы Asym

Eigen{Float64, Float64, Matrix{Float64}, Vector{Float64}}
values:
3-element Vector{Float64}:
  8.990924996526852
 24.03136729085444
 74.9777077126187
vectors:
3×3 Matrix{Float64}:
  0.595003  -0.178127   0.783736
 -0.660893   0.446485   0.603219
  0.457375   0.876882  -0.147937

👉 **Полученный нами результат вычисления представляет собой собственные значения и собственные векторы симметричной матрицы:**

- **Собственные значения (eigenvalues)**:
```julia
[8.990924996526852, 24.03136729085444, 74.9777077126187]
```
Собственные значения отражают, как исходная матрица масштабирует (растягивает или сжимает) пространство вдоль направлений, заданных соответствующими собственными векторами. 

Чем больше собственное значение, тем сильнее растяжение в направлении его собственного вектора.

<br>

- **Собственные векторы (eigenvectors)**:

Это столбцы (векторы) нашей матрицы –  направления (нормализованные векторы длины 1), по которым происходит масштабирование пространства. Каждый собственный вектор соответствует собственному значению с тем же индексом:

- Вектор 1 `[0.595003, -0.660893, 0.457375]` соответствует собственному значению `8.9909`.
  
- Вектор 2`[-0.178127, 0.446485, 0.876882]` соответствует собственному значению `24.0314`.
  
- Вектор 3 `[0.783736, 0.603219, -0.147937]` соответствует собственному значению `74.9777`.

<br>

>**Говоря простыми словами наша исходная матрица при умножении на данные векторы просто меняет их длину, не меняя направления.**
>**Эти три направления ортогональны (перпендикулярны друг другу), так как исходная матрица симметрична.**
>**Таким образом, полученный нами результат — не что инное как полное разложение матрицы на ортогональные направления и коэффициенты масштабирования.**

<br>

#### 🔹 Рассмотрим доступ к собственным значениям и векторам матрицы:


- **Задаем исходную матрицу A, которая будет использована для демонстрации собственных значений и векторов**

```julia
A = [
    25.0  15.0  -5.0;
    15.0  18.0   0.0;
    -5.0   0.0  11.0
]
```

A = [
    25.0  15.0  -5.0;
    15.0  18.0   0.0;
    -5.0   0.0  11.0
]

- **Создаём симметричную матрицу.**

Симметричная матрица важна для вычисления собственных значений и векторов, так как обеспечивает реальные значения и ортогональность собственных векторов.

In [49]:
Asym = A + A'

3×3 Matrix{Float64}:
  50.0  30.0  -10.0
  30.0  36.0    0.0
 -10.0   0.0   22.0

- **Выполняем EVD-разложение**

Вычисляем собственные значения и собственные векторы, получая EVD-разложение матрицы.

In [50]:
AsymEig = eigen(Asym)

Eigen{Float64, Float64, Matrix{Float64}, Vector{Float64}}
values:
3-element Vector{Float64}:
  8.990924996526852
 24.03136729085444
 74.9777077126187
vectors:
3×3 Matrix{Float64}:
  0.595003  -0.178127   0.783736
 -0.660893   0.446485   0.603219
  0.457375   0.876882  -0.147937

- **Получаем доступ к собственным значениям и векторам:**
  
  

**Собственные значения** - показывают степень масштабирования пространства по соответствующим направлениям (собственным векторам):

In [51]:
λ = AsymEig.values                     # Собственные значения
println("Собственные значения: ", λ)   # Вывод собственных значений

Собственные значения: 

[8.990924996526852, 24.03136729085444, 74.9777077126187]


**Собственные векторы** - задают направления, по которым происходит масштабирование:

In [52]:
V = AsymEig.vectors                    # Собственные векторы
println("Собственные векторы: \n", V)  # Вывод собственных векторов

Собственные векторы: 


[0.5950031751375643 -0.17812697674608458 0.7837359260181437; -0.660892505642328 0.446484526031047 0.6032185872473069; 0.4573754667251214 0.8768821746221844 -0.1479369266540509]


- **Проверим корректность факторизации**
  
А именно, что произведение `Asym` на `V` дает то же самое, что и произведение `V` на диагональную матрицу собственных значений

In [53]:
println("Проверка разложения: ", norm(Asym * V - V * Diagonal(λ)))  # должно быть ≈ 0

Проверка разложения: 4.8379234241467095e-14

<br>

<br>


#### 🔹 Использование EVD-факторизации для оптимизации вычислений.

**Используя EVD-факторизацию, обратная матрица вычисляется намного эффективнее, особенно для больших матриц**

In [54]:
Asym_inv_eig = inv(AsymEig)

3×3 Matrix{Float64}:
  0.0488889  -0.0407407   0.0222222
 -0.0407407   0.0617284  -0.0185185
  0.0222222  -0.0185185   0.0555556

**Проверим точность вычисления обратной матрицы :**

In [55]:
println("Эффективно вычисленная обратная матрица через EVD:\n", Asym_inv_eig)

Эффективно вычисленная обратная матрица через EVD:
[0.048888888888888864 -0.04074074074074075 0.022222222222222202; -0.04074074074074074 0.06172839506172842 -0.01851851851851851; 0.022222222222222185 -0.0185185185185185 0.055555555555555594]


**Убедимся в корректности (должна получиться единичная матрица)**

Этим шагом мы проверяем, что при умножении исходной матрицы на обратную, получается единичная матрица (I):

In [56]:
println("Проверка корректности вычисления обратной матрицы:\n", Asym_inv_eig * Asym)

Проверка корректности вычисления обратной матрицы:
[0.9999999999999988 -1.1102230246251565e-15 -2.220446049250313e-16; 8.881784197001252e-16 1.000000000000001 1.6653345369377348e-16; -1.7763568394002505e-15 -4.440892098500626e-16 1.000000000000001]


**Сравним со стандартным способом вычисления обратной матрицы, что менее эффективно. Но так ли это?**

Проверим, насколько сильно отличается результат, полученный через стандартный подход

In [57]:
Asym_inv_standard = inv(Asym)
println("Разница между двумя способами вычисления обратной матрицы: ", norm(Asym_inv_eig - Asym_inv_standard))

Разница между двумя способами вычисления обратной матрицы: 7.28583859910259e-17

#### 🔹 **Выводы:**

Использование специальных типов факторизаций в Julia позволяет:

👉 Хранить результаты вычислений компактно и логично.
  
👉 Быстро извлекать значения и векторы для дальнейших вычислений.
  
👉 Автоматически использовать оптимизированные методы, учитывающие структуру факторизации.


<br>

<br>

---

<br>

### 🟢 **2.6 Другие менее распространённые факторизации**

Помимо наиболее распространённых факторизаций, уже рассмотренных нами выше, существуют также менее известные, но важные методы разложения матриц, которые применяются в специфических и продвинутых задачах линейной алгебры и численного анализа.

<br>

### 📌 **Разложение Шура (Schur Decomposition)**

<br>

**Разложение Шура** представляет собой представление квадратной матрицы $A$ в виде произведения:

<br>

$$
A = Q \cdot T \cdot Q^*
$$

**где:**

- $Q$ — унитарная (ортогональная) матрица, то есть выполняется условие $Q^* Q = I$;
  
- $T$ — верхняя треугольная (для комплексной матрицы) или верхняя квазитреугольная матрица (для вещественной матрицы, известная как **Real Schur form**).

<br>


**Важно:**
>Однако, разработчики языка **Julia** приняли решение называть при разложении Шура унитарную матрицу в объекте не $Q$, а $Z$, чтобы избежать путаницы с другими разложениями, например **QR-разложением**, где уже используется буква $Q$. Таким образом, для ясности, это чисто техническое решение, сделанное разработчиками **Julia**.

<br>

#### 🔹 **Использование разложения Шура:**

- Удобно для вычисления собственных значений (на диагонали матрицы $T$).
  
- Применяется в теории управления и численном анализе.
  
- Эффективно используется при реализации QR-алгоритма нахождения собственных значений.

<br>

### 🗒  **Пример использования разложения Шура в Julia:**

#### 🔹 **Пример 1.**

- **Определяем матрицу $A$:**

```julia 
A = [
     4  3; 
    -5 -2
    ]
```

In [58]:
A = [4 3; -5 -2] # Создание матрицы A

2×2 Matrix{Int64}:
  4   3
 -5  -2

- **Выполняем разложение Шура (Schur decomposition) матрицы**:

In [59]:
SchurF = schur(A) # Получение матрицы Шура

Schur{Float64, Matrix{Float64}, Vector{ComplexF64}}
T factor:
2×2 Matrix{Float64}:
  1.0      0.837722
 -7.16228  1.0
Z factor:
2×2 Matrix{Float64}:
 0.811242  -0.58471
 0.58471    0.811242
eigenvalues:
2-element Vector{ComplexF64}:
 0.9999999999999999 + 2.449489742783178im
 0.9999999999999999 - 2.449489742783178im

**Получаем компоненты факторизации**

Мы помним, что разработчики языка Julia приняли решение называть унитарную матрицу в этом объекте не $Q$, а $Z$, чтобы избежать путаницы с другими разложениями.

- $Q$ – унитарная (ортогональная для вещественных матриц) матрица:

In [60]:
Q = SchurF.Z # Получение матрицы Q

2×2 Matrix{Float64}:
 0.811242  -0.58471
 0.58471    0.811242

- $𝑇$ – верхняя квазитреугольная матрица Шура (для вещественных матриц блоками 1x1 или 2x2)

In [61]:
T = SchurF.T # Получение верхнетреугольной матрицы T

2×2 Matrix{Float64}:
  1.0      0.837722
 -7.16228  1.0

👉 **Матрица $T$ имеет верхнетреугольную форму, что облегчает чтение собственных значений непосредственно с диагонали.**

<br>

---

<br>

### 📌 **Разложение Жордана (Jordan Normal Form)**

<br>

**Разложение Жордана** представляет собой представление квадратной матрицы $A$ в виде подобия с блочно-диагональной матрицей:

<br>

$$
A = P \cdot J \cdot P^{-1}
$$
**где:**

- $P$ - **Матрица перехода** - содержит в качестве столбцов собственные и обобщенные (присоединенные) векторы. Используется для перехода в новый базис, в котором исходная матрица имеет максимально простой вид (Жорданова форма).
  
- $J$ -**Матрица Жордана** - это верхнетреугольная матрица, содержащая собственные значения на диагонали и единицы над диагональю, если собственных векторов недостаточно для диагонализации.

<br>

>**Разложение применяется, когда матрица не диагонализируема, то есть ей не хватает полного набора собственных векторов.**

<br>

#### 🔹 **Использование разложения Жордана:**

- Теоретический анализ свойств матриц и операторов.
  
- Решение дифференциальных и разностных уравнений.
  
- Изучение структуры линейных операторов.

<br> 

📚 В Julia нет встроенной функции для прямого вычисления **разложения Жордана (Jordan decomposition)**, поскольку это разложение неустойчиво при численных вычислениях (чувствительно к погрешностям).

**Поэтому используют дополнительные пакеты:**
- Для численных расчетов лучше всего использовать:
   - `JordanForm.jl`
   - `LinearAlgebraX.jl`
- Для символьных расчетов рекомендуется:
   - `Symbolics.jl`
  
<br>

### 🗒  **Пример численного вычисления с помощью пакета `JordanForm.jl`**

#### 🔹 **Пример 1.**


**Установим пакет `JordanForm.jl`:**

In [62]:
using Pkg
Pkg.add("JordanForm")

   Resolving 

package versions...


  No Changes to `~/Introduction-to-Julia/Project.toml`
  No Changes to `~/Introduction-to-Julia/Manifest.toml`


- **Производим подключение пакета JordanForm для вычисления разложения Жордана**

In [63]:
using JordanForm

[ Info: Precompiling SymbolicsPreallocationToolsExt [d479e226-fb54-5ebe-a75e-a7af7f39127f] (cache misses: wrong dep version loaded (2))


ERROR: 

LoadError: 

UndefVarError: `ForwardDiff` not defined in `PreallocationTools`
Stacktrace:
 [1] include
   @ ./Base.jl:557 [inlined]
 [2] 

include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt128}}, source::Nothing)
   @ Base ./loading.jl:2881
 [3] top-level scope
   @ stdin:6
in expression starting at ~/.julia/packages/Symbolics/Eas9m/ext/SymbolicsPreallocationToolsExt.jl:1
in expression starting at stdin:6
┌ Error: Error during loading of extension SymbolicsPreallocationToolsExt of Symbolics, use `Base.retry_load_extensions()` to retry.
│   exception =
│    1-element ExceptionStack:
│    Failed to precompile SymbolicsPreallocationToolsExt [d479e226-fb54-5ebe-a75e-a7af7f39127f] to "~/.julia/compiled/v1.11/SymbolicsPreallocationToolsExt/jl_mmHMhm".
│    Stacktrace:
│      [1] error(s::String)
│        @ Base ./error.jl:35
│      [2] compilecache(pkg::Base.PkgId, path::String, internal_stderr::IO, internal_stdout::IO, keep_loaded_modules::Bool; flags::Cmd, cacheflags::Base.CacheFlags, reason

- **Задаем исходную матрицу $A$, для которой хотим найти Жорданову форму:**

```julia
A = [6 2 1; 
     0 3 0; 
     0 0 3]
```


In [64]:
A = [6 2 1; 0 3 0; 0 0 3] # Создание матрицы A

3×3 Matrix{Int64}:
 6  2  1
 0  3  0
 0  0  3

- **Вычисляем Жорданову нормальную форму $J$ и матрицу перехода $P$:**

In [65]:
J, P = jordan_form(A) # Получение жордановой формы матрицы A

JordanFactorization{ComplexF64, ComplexF64}(ComplexF64[-0.6666666666666666 + 0.0im -0.3333333333333333 + 0.0im 1.0 + 0.0im; 1.0 + 0.0im 0.0 + 0.0im -0.0 - 0.0im; 0.0 + 0.0im 1.0 + 0.0im -0.0 - 0.0im], ComplexF64[3.0 + 0.0im 0.0 + 0.0im 0.0 + 0.0im; 0.0 + 0.0im 3.0 + 0.0im 0.0 + 0.0im; 0.0 + 0.0im 0.0 + 0.0im 6.0 + 0.0im])

- **Выводим полученную Жорданову форму матрицы $A$ :**

In [66]:
println("Жорданова форма J:\n", J) # Вывод жордановой формы

Жорданова форма J:
ComplexF64

[-0.6666666666666666 + 0.0im -0.3333333333333333 + 0.0im 1.0 + 0.0im; 1.0 + 0.0im 0.0 + 0.0im -0.0 - 0.0im; 0.0 + 0.0im 1.0 + 0.0im -0.0 - 0.0im]


- **Выводим матрицу перехода (базиса), состоящую из собственных и присоединенных векторов :**

In [67]:
println("Матрица перехода P:\n", P) # Вывод матрицы перехода

Матрица перехода P:
ComplexF64

[3.0 + 0.0im 0.0 + 0.0im 0.0 + 0.0im; 0.0 + 0.0im 3.0 + 0.0im 0.0 + 0.0im; 0.0 + 0.0im 0.0 + 0.0im 6.0 + 0.0im]


- **Проверяем корректность полученного разложения, вычисляя разницу между исходной матрицей и произведением матриц разложения (должна быть очень близка к 0) :**

In [68]:
println("Проверка корректности разложения: ", norm(A - P*J*inv(P))) # должно быть ≈ 0

Проверка корректности разложения: 8.552127740444998

❌ **Полученный результат означает, что разложение некорректно, так как оно должно быть очень близким к нулю (например, порядка 1e-12 или меньше).**

Это связано с тем, что алгоритмы численного вычисления Жордановой формы являются неустойчивыми и могут давать существенные ошибки из-за погрешностей округления. Именно поэтому вычисление Жордановой формы практически всегда рекомендуют выполнять:

- либо аналитически (символьно),
  
- либо для матриц, которые специально подобраны и не дают ошибок округления.

<br>

**Как же исправить эту проблему?**

:) Используйте символьное вычисление !!!

<br>

### ✅ **Символьное вычисление с пакетом `Symbolics.jl`**



**Использование символьных вычислений**

- Позволяет нам получить аналитические формулы.
  
- Удобно для теоретических и исследовательских задач, когда числовых значений элементов нет или они неизвестны заранее.
  
- Упрощает решение общих задач линейной алгебры и матричного анализа.

<br>

**Установка:**

```julia
Pkg.add("Symbolics")
Pkg.add("SymbolicUtils")
```

#### 🔹 **Пример 2.**

Корректный пример с символикой:

**Выполним установку пакета `Symbolics.js` в Julia:**

In [69]:
using Pkg
Pkg.add("Symbolics")
Pkg.add("SymbolicUtils")

   Resolving package versions...


  No Changes to `~/Introduction-to-Julia/Project.toml`
  No Changes to `~/Introduction-to-Julia/Manifest.toml`
   Resolving package versions...


  No Changes to `~/Introduction-to-Julia/Project.toml`
  No Changes to `~/Introduction-to-Julia/Manifest.toml`


In [70]:
using Symbolics, LinearAlgebra

**Определяем матрицу $A$:**

In [71]:
A = [6 2 1; 0 3 0; 0 0 3] # Создание матрицы A

3×3 Matrix{Int64}:
 6  2  1
 0  3  0
 0  0  3

**Производим символьное вычисление собственных значений:**

Используем функцию `eigvals(A)`

In [72]:
eigvals_A = eigvals(A) # Вычисление собственных значений матрицы A

3-element Vector{Float64}:
 3.0
 3.0
 6.0

In [73]:
println("Собственные значения:\n", eigvals_A) # Вывод собственных значений

Собственные значения:
[3.0, 3.0, 6.0]


<br>

#### 🔹 **Пример 3.**

На этом примере продемонстрируем использование символьных вычислений в Julia с помощью пакета Symbolics.jl. Убедимся, что символьные вычисления помогают аналитически исследовать матрицы, позволяя работать не только с числовыми, но и с переменными элементами матриц.

**Мы зададим символьные переменные, которые затем можно использовать в символьной матрице. Это означает, что матрица может содержать не числа, а переменные, полезные при аналитических расчётах.**

1. Задаём **символьные переменные** (элементы матрицы):

In [74]:
@variables a11 a12 a13 a21 a22 a23 a31 a32 a33 # Создание символьных переменных

9-element Vector{Num}:
 a11
 a12
 a13
 a21
 a22
 a23
 a31
 a32
 a33

 2. Определение **символьной матрицы**, которая состоит полностью из символьных элементов. Её можно использовать для вывода общих формул и аналитического исследования свойств матриц, таких как собственные значения и векторы в общем виде.

In [75]:
A = [a11 a12 a13; 
     a21 a22 a23; 
     a31 a32 a33]

3×3 Matrix{Num}:
 a11  a12  a13
 a21  a22  a23
 a31  a32  a33

3. Получаем **характеристический полином** (аналитически!):

In [76]:
λ = Symbolics.scalarize(@variables λ)[1] # Создание символьной переменной λ
char_poly = det(A - λ * I(3)) # Вычисление характеристического полинома

(a11 - λ)*((a22 - λ)*(a33 - λ) - a23*a32) - a12*(a21*(a33 - λ) - a23*a31) + a13*(a21*a32 - (a22 - λ)*a31)

In [77]:
println("Характеристический полином матрицы A:\n", char_poly) # Вывод характеристического полинома

Характеристический полином матрицы A:
(

a11 - λ)*((a22 - λ)*(a33 - λ) - a23*a32) - a12*(a21*(a33 - λ) - a23*a31) + a13*(a21*a32 - (a22 - λ)*a31)


4. Затем приводится конкретная числовая матрица, для которой будут вычислены собственные значения. Из этого видно, как можно использовать символьные методы и подходы и для конкретных числовых матриц.

In [78]:
B = [6 2 1; 
     0 3 0; 
     0 0 3] # Создание матрицы B

3×3 Matrix{Int64}:
 6  2  1
 0  3  0
 0  0  3

5. Вычисляем **характеристический полином** конкретной матрицы $B$:

In [79]:
char_poly_B = det(B - λ * I(3)) # Вычисление характеристического полинома матрицы B

(6 - λ)*((3 - λ)^2)

In [80]:
println("Характеристический полином матрицы B:\n", char_poly_B) # Вывод характеристического полинома

Характеристический полином матрицы B:
(6 - λ)*((3 - λ)^2)


**Это даст вам точный символьный вид характеристического полинома:**

$$
\det(B - λ I) = (6 - λ)(3 - λ)^2
$$


Мы можем вывести также собственные значния матрицы $B$:

In [81]:
eigvals_B = eigvals(B)  # Вычисление собственных значений матрицы B
println("Собственные значения матрицы B:\n", eigvals_B) # Вывод собственных значений матрицы B


Собственные значения матрицы B:
[3.0, 3.0, 6.0]


👉 **На этом примере мы изучили:**

- Как использовать символьную матрицу и зачем её задавать.
  
- Как получить аналитические формулы (например, характеристический полином).
  
- Как комбинировать символьный и численный подходы для более глубокого анализа матриц.

<br>

<br>


### 📚 **Дополнительные возможности с `LinearAlgebraX.jl`**

`LinearAlgebraX.jl` - это другой мощный пакет для расширенных матричных вычислений.


#### 🔹 **Пример 4.**

**Выполним установку пакета `LinearAlgebraX.jl`**:

In [82]:
using Pkg
Pkg.add("LinearAlgebraX")

   Resolving package versions...


  No Changes to `~/Introduction-to-Julia/Project.toml`
  No Changes to `~/Introduction-to-Julia/Manifest.toml`


In [83]:
using LinearAlgebraX # Подключаем пакет LinearAlgebraX

**Определяем матрицу $A$:**

In [84]:
A = [6 2 1; 0 3 0; 0 0 3] # Создание матрицы A

3×3 Matrix{Int64}:
 6  2  1
 0  3  0
 0  0  3

**Произведём вычисление Жордановой формы:**

In [85]:
J = jordan_form(A) # Получение жордановой формы матрицы A

JordanFactorization{ComplexF64, ComplexF64}(ComplexF64[-0.6666666666666666 + 0.0im -0.3333333333333333 + 0.0im 1.0 + 0.0im; 1.0 + 0.0im 0.0 + 0.0im -0.0 - 0.0im; 0.0 + 0.0im 1.0 + 0.0im -0.0 - 0.0im], ComplexF64[3.0 + 0.0im 0.0 + 0.0im 0.0 + 0.0im; 0.0 + 0.0im 3.0 + 0.0im 0.0 + 0.0im; 0.0 + 0.0im 0.0 + 0.0im 6.0 + 0.0im])

In [86]:
println("Жорданова форма (J):\n", J) # Вывод жордановой формы

Жорданова форма (J):
JordanFactorization

{ComplexF64, ComplexF64}(ComplexF64[-0.6666666666666666 + 0.0im -0.3333333333333333 + 0.0im 1.0 + 0.0im; 1.0 + 0.0im 0.0 + 0.0im -0.0 - 0.0im; 0.0 + 0.0im 1.0 + 0.0im -0.0 - 0.0im], ComplexF64[

3.0 + 0.0im 0.0 + 0.0im 0.0 + 0.0im; 0.0 + 0.0im 3.0 + 0.0im 0.0 + 0.0im; 0.0 + 0.0im 0.0 + 0.0im 6.0 + 0.0im])


---

<br>

#### 📝 **Заключительные выводы:**

**Факторизация** это мощный инструмент в вычислительной линейной алгебре, делающий сложные вычислительные задачи более простыми и точными. Разные типы факторизаций решают разные задачи, и выбор конкретного подхода зависит от типа исходной матрицы и задачи, которую нужно решить.


⚙️ **Как выбрать нужный тип факторизации?**

Выбор зависит от конкретной задачи и свойств матрицы:

| Свойство матрицы       | Подходящая факторизация        |
|------------------------|--------------------------------|
| Общая матрица          | LU, QR, SVD                    |
| Симметричная положительно определённая | Cholesky, LU, SVD, EVD |
| Неквадратная           | QR, SVD                        |
| Диагонализируемая      | EVD                            |

**Быстрые алгоритмы по сложности** (от простого к сложному):

- **Cholesky** (~$n^3/3$)
  
- **LU** (~$2n^3/3$)
  
- **QR** (~$2n^3$)
  
- **SVD** (~$4n^3$) — наиболее затратный, но универсальный и устойчивый.
  

**Другие менее распространённые факторизации:**

- **Разложение Шура** (Schur Decomposition).
  
- **Разложение Йордана** (Jordan Normal Form).

Эти менее распространённые факторизации широко используются в теоретических исследованиях, а также при решении специфических задач в математике, физике, теории управления и других областях, где необходимы глубокие структурные свойства матриц.

<br>

<br>

---

<br>

<br>

## 🧑‍💻 **3. Специальные матричные структуры.**

<br>

Традиционно считается, что **структура матриц** играет ключевую роль в линейной алгебре.

👉 Чтобы наглядно понять, насколько это важно, рассмотрим систему линейных уравнений с большой матрицей (например, размером 1000×1000):

In [87]:
n = 1000         # Размер матрицы
A = randn(n, n)  # Создание случайной матрицы

1000×1000 Matrix{Float64}:
  0.554895    1.16938     0.362124   …   1.66296   -0.778725   1.30059
  0.536264   -1.34889    -0.572647      -1.71964    1.28349    0.40343
 -0.403118   -0.694      -0.731242       1.97437    0.337213  -0.695954
 -0.321995    1.11497     0.563094       1.46769   -0.832411   1.31266
 -0.679623   -1.31703     0.152239      -0.200897   0.729268  -0.0447025
 -0.202479    0.784475    0.767135   …   1.40981    2.9838    -0.788995
 -0.278266   -0.514358    1.11249       -1.0783     0.640713  -1.84387
 -0.0742406   0.0764335  -1.14225       -0.218901  -0.410812  -0.70121
  0.510634    0.0408098  -0.0535482      1.93524    2.2844    -1.9413
  0.195141   -1.60993     2.0477        -1.09366    0.694365   0.361395
  0.268819    1.58525    -1.44925    …  -0.826922   0.760806   1.26464
  0.0998014  -0.821223   -0.225563       1.06294    0.176989   0.677608
 -1.21727     0.417799   -0.951277      -0.62466    0.812384  -0.343996
  ⋮                                  ⋱      

**Julia позволяет проверять свойства матриц, автоматически используя их структуры.**

Например, проверим симметричность матрицы:

In [88]:
A_sym = A + A'  # Создание симметричной матрицы

issymmetric(A_sym)  # true

true

**Однако иногда численные погрешности могут мешать корректному определению структуры матрицы:**

In [89]:
A_sym_noisy = copy(A_sym)   # Создаем копию матрицы A_sym
A_sym_noisy[1, 2] += 5eps() # Добавляем шум к элементу матрицы

issymmetric(A_sym_noisy)    # false из-за численной погрешности

false

**К счастью, мы можем явно указать структуру матрицы, используя встроенные типы Julia, такие как:**

- `Diagonal` (диагональная),
  
- `Triangular` (треугольная),
  
- `Symmetric` (симметричная),
  
- `Hermitian` (эрмитова),
  
- `Tridiagonal` (трёхдиагональная),
  
- `SymTridiagonal` (симметричная трёхдиагональная).

Например:

In [90]:
A_sym_explicit = Symmetric(A_sym_noisy) # Явное приведение к симметричной матрице

1000×1000 Symmetric{Float64, Matrix{Float64}}:
  1.10979     1.70564   -0.0409944  …   2.13945     0.270355    1.92895
  1.70564    -2.69779   -1.26665       -1.02602     1.98062     1.07846
 -0.0409944  -1.26665   -1.46248        1.82691     0.885346   -2.32123
  0.0490773  -0.240169  -0.275155       0.54541    -0.810998    1.54001
  0.649251   -0.938494   2.10015        0.363362    0.378707    0.109233
 -2.18434     1.34603    0.269377   …   2.37141     3.28653     0.359871
 -0.417484   -0.247889  -0.0789008     -0.0949465   0.903062   -1.96227
 -1.15647    -0.543741  -1.25623        1.55703    -0.391147   -3.03795
  1.91835    -1.61578   -0.593558       1.32224     1.43117    -1.814
  0.673259   -2.0138     1.2299        -0.388771    1.67766    -0.932041
 -0.795257    2.08723   -1.08265    …   0.456155    0.805937    0.940593
 -0.0880234  -0.896981  -1.65498        0.705519   -0.0635017   1.43605
 -0.851169    0.720895  -1.56864       -1.44336     0.56192    -0.0235128
  ⋮          

**Сравним скорость вычисления собственных значений для разных случаев:**

Использование типа `Symmetric` сделает вычисления примерно в 5 раз быстрее!

In [91]:
@time eigvals(A_sym);             # без явного указания структуры
@time eigvals(A_sym_noisy);       # численная погрешность
@time eigvals(A_sym_explicit);    # с явно указанной структурой

  0.236851 seconds (50 allocations: 7.990 MiB, 2.79% compilation time)
  2.183225 seconds (27 allocations: 7.928 MiB)


  0.336254 seconds (133.38 k allocations: 14.910 MiB, 43.40% compilation time)


<br>

#### 🚩 **Большая проблема и её решение**

Использование специальных типов, таких как `Tridiagonal` и `SymTridiagonal`, позволяет работать с очень большими трёхдиагональными матрицами, которые было бы невозможно хранить и вычислять как обычные (плотные) матрицы.

<br>

**Например, следующую задачу невозможно было бы решить на обычном ноутбуке с плотной матрицей:**

In [92]:
n = 1_000_000    # Размер матрицы
A_large = SymTridiagonal(randn(n), randn(n - 1)) # Создание большой трехдиагональной матрицы

1000000×1000000 SymTridiagonal{Float64, Vector{Float64}}:
 -0.827933  -0.104947    ⋅        …    ⋅          ⋅          ⋅ 
 -0.104947   1.9722    -0.783711       ⋅          ⋅          ⋅ 
   ⋅        -0.783711   0.673113       ⋅          ⋅          ⋅ 
   ⋅          ⋅        -1.14982        ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅             ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅        …    ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅             ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅             ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅             ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅             ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅        …    ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅             ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅             ⋅          ⋅          ⋅ 
  ⋮                               ⋱                        
   ⋅          ⋅          ⋅             ⋅          

In [93]:
@time eigmax(A_large) # Вычисление максимального собственного значения

  1.680972 seconds (616.82 k allocations: 215.854 MiB, 33.89% gc time, 18.99% compilation time)


6.4873265805337255

👉 **Благодаря специальным структурам матриц можно эффективно решать масштабные численные задачи, которые иначе были бы слишком ресурсоёмкими для обычных компьютеров.**

<br>

---

<br>

<br>

## ✨ **4. Общая линейная алгебра (Generic linear algebra)**

<br>

Стандартный подход к реализации численной линейной алгебры заключается в использовании подпрограмм из библиотек BLAS и LAPACK. Julia также придерживается этого подхода для матриц с элементами типов `Float32`, `Float64`, `Complex{Float32}` или `Complex{Float64}`.

Однако Julia дополнительно поддерживает **обобщённую (generic)** линейную алгебру. Это позволяет работать с матрицами и векторами, элементами которых являются, например, рациональные числа, комплексные числа произвольной точности и другие типы данных, не обязательно числовые.

<br>

### 🔢 **Рациональные числа (Rational numbers)**

Julia поддерживает рациональные числа из коробки. Рациональные числа можно создавать с помощью двойного слэша `//`:


In [94]:
r = 1//2 # Рациональное число

1//2

**Рациональные числа представляются в виде дроби, что позволяет сохранять точность вычислений, избегая ошибок округления, характерных для чисел с плавающей точкой.**

#### 🔹 **Пример: решение рациональной системы линейных уравнений**

Рассмотрим пример решения системы линейных уравнений, элементы которой — рациональные числа. Чтобы избежать проблем переполнения, характерных при работе с рациональными числами, будем использовать типы большой точности `BigInt`:

- **Создадим матрицу с рациональными элементами типа `BigInt` :**
  
  Генерация случайной рациональной матрицы 3x3

In [95]:
Arational = Matrix{Rational{BigInt}}(rand(1:10, 3, 3)) // 10

3×3 Matrix{Rational{BigInt}}:
 3//10   1     1//10
 1//10  2//5   4//5
 4//5   3//10  4//5

- **Создаём вектор $x$, заполненный единицами :**

In [96]:
x = fill(1, 3)

3-element Vector{Int64}:
 1
 1
 1

- **Вычисляем вектор правой части системы :**

In [97]:
b = Arational * x

3-element Vector{Rational{BigInt}}:
  7//5
 13//10
 19//10

- **Решаем систему A * solution = b :**

In [98]:
solution = Arational \ b

3-element Vector{Rational{BigInt}}:
 1
 1
 1

- **Выводим результаты :**

In [99]:
println("Матрица A:\n", Arational)
println("\nВектор b:\n", b)
println("\nРешение системы Ax = b:\n", solution)

Матрица A:
Rational{BigInt}

[3//10 1 1//10; 1//10 2//5 4//5; 4//5 3//10 4//5]

Вектор b:
Rational{BigInt}

[7//5, 13//10, 19//10]

Решение системы Ax = b:
Rational{BigInt}[1, 1, 1]


- **Проверка корректности :**

In [100]:
println("\nПроверка: A * solution == b? ", Arational * solution == b) # true


Проверка: A * solution == b? true

🔴 **Мы можем также явно использовать LU-факторизацию, что полезно при многократном решении системы с одной и той же матрицей :**

Наша **LU-факторизация** матрицы с рациональными элементами будет выглядеть так:

In [101]:
Arational_lu = lu(Arational)

LU{Rational{BigInt}, Matrix{Rational{BigInt}}, Vector{Int64}}
L factor:
3×3 Matrix{Rational{BigInt}}:
  1      0     0
 3//8    1     0
 1//8  29//71  1
U factor:
3×3 Matrix{Rational{BigInt}}:
 4//5   3//10    4//5
  0    71//80   -1//5
  0      0     111//142

- **Решение с использованием факторизации :**

In [102]:
solution_lu = Arational_lu \ b

3-element Vector{Rational{BigInt}}:
 1
 1
 1

✅ **Оба способа сохраняют точность без перехода к типам с плавающей точкой.**

<br>

#### 🚩 **Преимущества обобщённой линейной алгебры:**

- Возможность работать с широким классом числовых типов, включая точные рациональные и комплексные числа произвольной точности.
  
- Исключение ошибок округления и повышение точности вычислений в теоретических и аналитических задачах.
  
- Удобство и гибкость при работе с математическими и численными задачами различной сложности.

<br>

<br>

---

<br>

### **Упражнения** 🚀 

<br>

#### ✅ Здание 13.1 

Найдите собственные значения матрицы `𝐴`

```julia
A =
[
 140   97   74  168  131
  97  106   89  131   36
  74   89  152  144   71
 168  131  144   54  142
 131   36   71  142   36
]
```

Присвойте результат переменной `A_eigv`.

In [103]:
using LinearAlgebra

In [104]:
# Ваше решение:


In [ ]:
@assert A_eigv ==  [-128.49322764802145, -55.887784553056875, 42.7521672793189, 87.16111477514521, 542.4677301466143]

<br>

#### ✅ Здание 13.2 

Создайте диагональную матрицу, состоящую из собственных значений матрицы `A`.

Присвойте результат переменной `A_diag`.

In [106]:
# Ваше решение:


In [ ]:
@assert A_diag ==  [-128.493    0.0      0.0      0.0       0.0;
    0.0    -55.8878   0.0      0.0       0.0;
    0.0      0.0     42.7522   0.0       0.0;
    0.0      0.0      0.0     87.1611    0.0;
    0.0 0.0      0.0      0.0     542.468]

<br>

#### ✅ Здание 13.3

Создайте нижнюю треугольную матрицу (матрицу, в которой все элементы выше главной диагонали равны нулю) на основе матрицы `A` и сохраните результат в переменной `A_lowertri`.

In [108]:
# Ваше решение:


In [ ]:
@assert A_lowertri ==  [140    0    0    0   0;
  97  106    0    0   0;
  74   89  152    0   0;
 168  131  144   54   0;
 131   36   71  142  36]